Fixing complex objects checks for fixed children

mint-thompson opened the pull request:
Fixes #560 and completes task [CIMPL-496](https://standardhealthrecord.atlassian.net/browse/CIMPL-496).

When fixing a complex object to an element, children of that element or of that element's slicedElement may already have fixed values. When this happens, make sure that the two values do not conflict with one another.

The case of "an instance is assigned to a slice (such as `address[RecentAddress]`), and a later fixed value on a non-slice (such as `address.period.start`) would violate that slice" is out of scope for this task.

ngfreiter commented on src/fhirtypes/ElementDefinition.ts:
```diff
@@ -1300,6 +1311,33 @@ export class ElementDefinition {
+      // find the element on fhirValue that would get assigned to the child
+      const childPath = child.id.replace(`${this.id}.`, '').split('.');
+      let newChildValue = fhirValue;
+      for (const pathPart of childPath) {
+        if (newChildValue != null) {
+          newChildValue = newChildValue[pathPart];
+        }
```
I think this causes an issue in the case of an array. Consider this example:
```
Profile: MyObs
Parent: Observation
* valueCodeableConcept.coding.code = #bar
* valueCodeableConcept = #foo
```
This is pretty much the same situation as your test on `valueQuantity`, but because `valueCodeableConcept` contains an array, the test does not seem to work. You end up with `patternCode: "bar"` on `valueCodeableConcept.coding.code`, and `patternCodeableConcept: {coding: [{code: "foo"}]}` on `valueCodeableConcept`, and these two rules conflict.

The `fixedByAnyParent` function checks if a value is fixed by any parent of an element, and returns that element if so, and it does work with arrays, so my first thought was that you could use that function, but of course that is an issue because you can't use that function until you actually do fix the value on the parent element, but we are trying to decide if it is valid to fix the value. So one possibility would be to fix the value on the parent, determine if that fixing violated anything on the child elements, and un-fix the value on the parent if so. But maybe it is just simpler to add array handling logic here.

mint-thompson commented:
Thank you for providing the additional sample case! I've added a test case with that scenario, and I've updated `checkChildFixedValue` to traverse values that contain arrays.
